lib/rouge/lexer.rb

```diff
@@ -418,7 +418,7 @@ def reset!
     # Given a string, yield [token, chunk] pairs.  If no block is given,
     # an enumerator is returned.
     def lex(string, opts=nil, &b)
-      if opts
+      if opts && opts[:continue]
```

Comment:
This won't be enough, will it? The `opts` hash is no longer passed to `enum_for` later on and so these options will just silently disappear, won't they? Better than crashing, admittedly, but still not great.

Reply:
I didn't trace what happens to these options after this point. So perhaps TBD. What I do know is that going down this code path when :continue is not set is disastrous, so we need to stop the bleeding. Right now, 3.4.0 is unusable because of this.

Reply:
Yeah, my goal is to have 3.4.1 out tonight to fix it. I'm trying to work through the possible code paths to better understand the implications of any further changes.

Reply:
I'd also just add that, at this point, I'm leaning towards largely reverting to the previous behaviour, maybe with a deprecation warning thrown in.

This was intended as a performance optimisation, so returning to the status quo while we work out a more long term solution seems to me to be the most stable fix for the time being.

Reply:
I looked more into this and you're correct, you can't pass options that way anymore. That actually means I need to update my code.

For sure, this is a backwards compatibility issue. So checking for opts[:continue] explicitly before taking the codepath in question, then perhaps warning if other options are passed (assuming it's non-empty).

I can (and should) update Asciidoctor to pass the options through the constructor instead of through the lex method since the latter no longer works anyway.

Reply:
> 3.4.0 is unusable because of this.

It's probably fair to say that it's not backwards compatible. It's usable if options are passed to the Lexer constructor (or find_fancy).

Reply:
I've submitted a PR to your PR with some minor tweaks :P Having looked at this more closely, `opts` is all but ignored by `Lexer.lex`. Accordingly, I don't think it's necessary to fully revert to the 3.3.0 behaviour.

Reply:
I'll repeat what I wrote there:

> I still think this is too aggressive. Rouge 3.4.0 deprecates an API, drops the existing behavior, and warns all in the same release without any migration path. I think this whole change warrants a major release, so it really should be rolled back entirely until then. Users of the API need some sort of migration path at least.

Reply:
As it stands, I'm also going to be fielding bug reports in Asciidoctor until I put out new releases of both Asciidoctor and Asciidoctor PDF to stop using this API.

And for the record, this is an API that works fine in Pygments. So we're departing from that familiar ground as well. In Pygments, you find a lexer, then pass options when you use it.